src/org/jitsi/impl/neomedia/transform/fec/FlexFec03Receiver.java

```diff
@@ -0,0 +1,372 @@
+         * @param payloadLength the length of the source's payload
+         */
+        private void xorPayloads(RawPacket source, int sourceOffset,
+                                 RawPacket dest, int destOffset,
+                                 int payloadLength)
+        {
+            for (int i = 0; i < payloadLength; ++i)
+            {
+                dest.getBuffer()[destOffset + i] ^=
+                    source.getBuffer()[sourceOffset + i];
+            }
+        }
+
+        /**
+         * Do the final work when recovering an RTP packet (set the RTP version,
+         * the length, the sequence number, and the ssrc)
+         * @param fecPacket the fec packet
+         * @param recoveredPacket the media packet which was recovered
+         */
+        private boolean finishPacketRecovery(FlexFec03Packet fecPacket, RawPacket recoveredPacket)
+        {
+            // Set the RTP version to 2.
+            recoveredPacket.getBuffer()[0] |= 0x80; // Set the 1st bit
+            recoveredPacket.getBuffer()[0] &= 0xbf; // Clear the second bit
+
+            // Recover the packet length, from temporary location.
+            int length = RTPUtils.readUint16AsInt(recoveredPacket.getBuffer(), 2);
+            if (length > recoveredPacket.getBuffer().length)
+            {
+                logger.error("Length field of recovered packet is larger" +
+                    " than its buffer");
+                return false;
+            }
+            // The length field used in the xor does not include the header
+            // length, but we want to include the fixed header length when
+            // setting the length on the packet object
+            recoveredPacket.setLength(length + RawPacket.FIXED_HEADER_SIZE);
+
+            // Set the SN field.
+            RTPUtils.writeShort(recoveredPacket.getBuffer(), 2,
+                (short)missingSequenceNumber);
+
+            // Set the SSRC field.
+            RTPUtils.writeInt(recoveredPacket.getBuffer(), 8,
+                (int)fecPacket.getProtectedSsrc());
+
+            return true;
+        }
+
+        /**
+         * Recover a media packet based on the given FlexFEC packet and
+         * received media packets
+         * @return the recovered packet if successful, null otherwise
+         */
+        private RawPacket recover()
+        {
+            if (!canRecover())
+            {
+                return null;
+            }
+
+            //TODO: use a pool?
+            byte[] buf = new byte[1500];
+            RawPacket recoveredPacket = new RawPacket(buf, 0, 1500);
+            if (!startPacketRecovery(this.fecPacket, recoveredPacket))
+            {
+                return null;
+            }
+            for (Integer protectedSeqNum : fecPacket.getProtectedSequenceNumbers())
+            {
+                if (protectedSeqNum != missingSequenceNumber)
+                {
+                    RawPacket mediaPacket = mediaPackets.get(protectedSeqNum);
```

Comment:
Should we deal with the case of the media packet being cleared from the cache? Or leave it for when we switch to reusing the MediaStream's cache?

Reply:
hmm, i'm trying to think if this can actually happen.  `AbstractFECReceiver#reverseTransform` gets called, which saves the current fec/media packets.  It then calls `FlexFec03Receiver#doReverseTransform` which looks through the media and fec packets to see if recovery is possible, this is all done synchronously (and `AbstractFECReceiver#reverseTransform` itself is synchronized) so by the time we get here we know that all the media packets we need will be in `mediaPacketss`.

Reply:
Are we still planning to eventually remove the cache and re-use the existing one, though?

Reply:
yeah, it seems to me like that would be worth looking into